// @lafourchette/eslint-config-thefork/index.js
'use strict';

module.exports = {
  plugins: ['prettier'],
  extends: ['eslint-config-airbnb-base', 'plugin:prettier/recommended'],
  env: {
    es6: true,
    node: true,
  },
  rules: {
    'prettier/prettier': 'error',
    curly: ['error', 'all'],
    'no-param-reassign': ['error', { props: false }], // useful for .reduce()
    strict: 0,
    'func-names': [1, 'as-needed'],
    'max-len': [
      2,
      120,
      2,
      {
        ignoreUrls: true,
        ignoreComments: true,
        ignoreRegExpLiterals: true,
        ignoreStrings: true,
        ignoreTemplateLiterals: true,
      },
    ],
    'no-underscore-dangle': 0,
    // Re-enable ForOfStatement (disabled in eslint-config-airbnb-base)
    'no-restricted-syntax': [
      'error',
      {
        selector: 'ForInStatement',
        message:
          'for..in loops iterate over the entire prototype chain, which is virtually never what you want. Use Object.{keys,values,entries}, and iterate over the resulting array.',
      },
      {
        selector: 'LabeledStatement',
        message: 'Labels are a form of GOTO; using them makes code confusing and hard to maintain and understand.',
      },
      {
        selector: 'WithStatement',
        message: '`with` is disallowed in strict mode because it makes code impossible to predict and optimize.',
      },
    ],
  },
};
